Source/WebCore/platform/graphics/SourceBufferPrivate.cpp

```diff
@@ -614,27 +613,111 @@ bool SourceBufferPrivate::validateInitializationSegment(const SourceBufferPrivat
     return true;
 }
 
-void SourceBufferPrivate::didReceiveSampleForTrackId(uint64_t, Ref<MediaSample>&& mediaSample)
+void SourceBufferPrivate::didReceiveSample(Ref<MediaSample>&& sample)
+{
+    if (!isAttached())
+        return;
+
+    if (m_pendingAppendOperations.isEmpty())
+        m_pendingAppendOperations.append({ { }, { } });
+
+    DEBUG_LOG(LOGIDENTIFIER, sample.get());
+    m_pendingAppendOperations.last().mediaSamples.append(WTFMove(sample));
+}
+
+void SourceBufferPrivate::append(Ref<SharedBuffer>&& buffer)
 {
-    didReceiveSample(WTFMove(mediaSample));
+    append(buffer->extractData());
 }
 
-void SourceBufferPrivate::didReceiveSample(Ref<MediaSample>&& originalSample)
+void SourceBufferPrivate::append(Vector<unsigned char>&&)
 {
-    if (!m_isAttached || m_didReceiveSampleErrored) {
-        m_pendingMediaSamples.clear();
-        return;
+    RELEASE_ASSERT_NOT_REACHED();
+}
+
+void SourceBufferPrivate::processPendingOperations()
+{
+    while (!m_pendingAppendOperations.isEmpty()) {
+        auto& operation = m_pendingAppendOperations.first();
+        if (operation.initOperation) {
+            processInitSegment();
+            if (m_didReceiveInitializationSegmentErrored)
+                break;
+            if (operation.initOperation)
+                return; // will continue once Init completionHandler is run.
+        }
+        processMediaSamples();
+        if (m_didReceiveSampleErrored)
+            break;
+        m_pendingAppendOperations.removeFirst();
```

Comment:
Since we are removing the operation here, would it clarify things if we were taking the operation as the first step, and maybe pass it to processInitSegment or processMediaSamples?

Reply:
The initOperation isn't to be dropped until the init CompletionHandler as completed, Right now, the existence of the initOperation indicates that it hasn't completed yet. We would need to add a boolean now. 
I'll see what I can do, but I'm not sure it will end up being any better.